src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/DoNotGuardDictionaryRemoveByContainsKey.cs

```diff
@@ -0,0 +1,119 @@
+
+        public const string AdditionalDocumentLocationInfoSeparator = ";;";
+
+        public const string ConditionalOperation = nameof(ConditionalOperation);
+        public const string ChildStatementOperation = nameof(ChildStatementOperation);
+
+        internal static readonly DiagnosticDescriptor Rule = DiagnosticDescriptorHelper.Create(
+            RuleId,
+            s_localizableTitle,
+            s_localizableMessage,
+            DiagnosticCategory.Performance,
+            RuleLevel.IdeSuggestion,
+            s_localizableDescription,
+            isPortedFxCopRule: false,
+            isDataflowRule: false);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            context.EnableConcurrentExecution();
+            context.RegisterCompilationStartAction(OnCompilationStart);
+        }
+
+        private static void OnCompilationStart(CompilationStartAnalysisContext context)
+        {
+            var compilation = context.Compilation;
+
+            if (!compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemCollectionsGenericDictionary2, out var dictionaryType))
+                return;
+
+            context.RegisterOperationAction(AnalyzeOperation, OperationKind.Invocation);
+
+            static void AnalyzeOperation(OperationAnalysisContext context)
+            {
+                var invocationOperation = (IInvocationOperation)context.Operation;
+
+                if (invocationOperation.TargetMethod.Name != "ContainsKey")
+                    return;
+
+                if (invocationOperation.Parent is not IConditionalOperation parentConditionalOperation)
+                    return;
+
+                // we only want to report this diagnostic if the Contains/Remove pair is all there is
+
+                if (parentConditionalOperation.WhenFalse == null &&
+                    parentConditionalOperation.WhenTrue.Children.HasExactly(1))
+                {
+                    var properties = ImmutableDictionary.CreateBuilder<string, string>();
+                    properties[ConditionalOperation] = CreateLocationInfo(parentConditionalOperation.Syntax);
+
+                    switch (parentConditionalOperation.WhenTrue.Children.First())
+                    {
+                        case IInvocationOperation childInvocationOperation:
+                            if (childInvocationOperation.TargetMethod.Name == "Remove")
+                            {
+                                properties[ChildStatementOperation] = CreateLocationInfo(childInvocationOperation.Syntax.Parent);
+
+                                context.ReportDiagnostic(Diagnostic.Create(Rule, invocationOperation.Syntax.GetLocation(), properties.ToImmutable()));
```

Comment:
Please use the following syntax
```suggestion
                                context.ReportDiagnostic(invocationOperation.CreateDiagnostic(Rule, properties.ToImmutable()));
```